### ISSUE TITLE
Implement message based distributed architecture

### ISSUE THREAD

**Issue by kaustavdm** [muktakosh/unicorn#2]
- [x] Implement a `messages` module
- [x] Implement message serialization: Structs -> JSON -> Structs
- [x] Implement connectivity layer (using `nanomsg`)

**Comment by nootanghimire** [comment]
Just getting started on Rust and ZeroMQ basics. By the way, which Messaging pattern are we supposed to use? I'd guess Client/Server ?

**Comment by kaustavdm** [comment]
@nootanghimire Yes, client/server.

`core` acts as a server. Other components connect to the core to exchange messages.

Have fun!

**Comment by kaustavdm** [comment]
**Quick update:** I decided to write a network layer, replacing ZeroMQ.

Don't get me wrong. I love ZeroMQ. But, I realized that it is easier to write a low level networking implementation in Rust, than use ZeroMQ using a `libc` binding. It will take more work, but it will give better control and reliability.

**Comment by sankha93** [comment]
I just noticed that you replaced it with your own network layer, that means instead of messages we now have streams. So does that mean you'll define your own message format over the stream or change the message queue idea to stream based in the core?

**Comment by kaustavdm** [comment]
@sankha93 Each component will connect to each other through network streams and exchange messages. The `network` module will provide the transport. The `messages` module will provide the format and utilities for encoding and decoding.

Initially, I wanted to use zeromq because of the abstractions it provides over TCP. Even in zeromq, you need to define your own message formats, which get serialized and sent and received as bytes. But the existing rust wrappers for zeromq were not feature complete and did not have active development. Instead of going through the entire libzmq API, I thought of trying my hands on `std::net` to see if I can write a network layer.

This decision implies more work to be done, but in the end, we can come up a lean and focused network layer that does only the things that `unicorn` needs. In the process, it removes an external system dependency. For a project like unicorn, it is very essential to keep external dependencies as close to zero as possible

**Comment by kaustavdm** [comment]
Update: Had to add a couple of dependencies, but the new [network layer](https://github.com/muktakosh/unicorn/tree/master/src/network) is ready (with some [tests](https://github.com/muktakosh/unicorn/tree/master/tests/network.rs)).

**Comment by kaustavdm** [comment]
...Aaaaanndddd it's gone. That `network` layer that I had started writing has metamorphosed into a little something that I have named [`rpc`](https://github.com/muktakosh/unicorn/blob/master/src/rpc.rs). It is a neat wrapper around `nanomsg.rs`. Preliminary examination revealed that it is capable of distributing the architecture very smoothly.

I'll miss this joyride of experimenting (creating new stuff) and evaluating (seeing if it can be bent and ditching it if it does not) once `unicorn` has a stable API. Till then, I shall have fun.
